runtime/src/iree/hal/drivers/cuda/cuda_driver.c

```diff
@@ -70,12 +90,31 @@ static iree_status_t iree_hal_cuda_driver_create_internal(
   iree_status_t status =
       iree_hal_cuda_dynamic_symbols_initialize(host_allocator, &driver->syms);
 
-  // Initialize NCCL too if a channel provider is defined or any default
-  // collective group values.
+  int comm_world_size = 0;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_cuda_get_mpi_comm_world_size_from_env(&comm_world_size);
+  }
+
+  // Initialize NCCL too if MPI is used or default_count is set.
   if (iree_status_is_ok(status) &&
-      default_params->channel_provider.query_group_params) {
+      (comm_world_size > 0 || options->default_count > 0)) {
     status = iree_hal_cuda_nccl_dynamic_symbols_initialize(host_allocator,
                                                            &driver->syms);
+    if (comm_world_size > 0) {
+      status = iree_hal_mpi_dynamic_symbols_initialize(host_allocator,
```

Comment:
iree_hal_cuda_nccl_dynamic_symbols_initialize may have failed, in which case this would ignore that and overwrite the status. add a check for iree_status_is_ok &&